components/prism-factor.js

```diff
@@ -0,0 +1,277 @@
+(function (prism) {
+
+	var factor = {
+
+		comment: { // ! single-line exclamation point comments with whitespace after/around the !
+			pattern: /(^|\s|(?!"\s*))(?:![\t ].*|!$)/,
```

Comment:
This fixes a problem where strings containing `!` would be tokenised as comments starting before the string is closed.

Removing the lookahead in question, the test fails like:

```
: expected [ Array(23) ] to deeply equal [ Array(21) ]
      + expected - actual

           "string"
           "\" ! \""
         ]
         [
      -    "conventionally-named-word"
      -    "\"!"
      -  ]
      -  [
           "string"
      -    "\"\r\n\r\n\""
      +    "\"! \""
         ]
      -  "\\"
         [
           "string"
      -    "\"\""
      +    "\"\\\"\""
         ]
         [
           "string"
           "\"'\""
```

It's hard to exactly understand what goes wrong here (it's related to the CRLF because it doesn't happen on localhost/test.html), but this lookahead fixes it and makes it behave correctly regardless of line endings, which is the point.

Reply:
The test failed because you didn't just remove the lookahead.
Because the lookahead will always accept (aka. the pattern `"\s*`) cannot be matched, it's unnecessary and can be removed. So we get:

`(^|\s|)(?:![\t ].*|!$)` (the alternative is still there)

If the empty alternative is removed, the tests fail because the language of the RE changed, so let's keep it. But that means that our lookbehind can match the empty and will therefore always accept, so it can be removed.

```js
comment: {
	pattern: /(?:![\t ].*|!$)/,
	inside: ...
}
```

This will also pass all tests because the meaning of the pattern didn't change. But that's not what you intended, right?

---

So, why do the tests fail with: `(^|\s)(?:![\t ].*|!$)` ?

The simplest example which fails is:

```
"! "
```

In this example, `"!` is tokenized as `conventionally-named-word` because it matches the `[^!\r\n\t ]+!` part of the pattern. So what Prism does is the following (simplified):

1. It tries to match `comment` and fails because `"` doesn't match the lookbehind (`(^|\s)`).
2. It tries to match `conventionally-named-word` and succeeds. The text is now tokenized as `[["conventionally-named-word", "\"!"], " \""]`.
3. Prism tries to match string and fails because the `conventionally-named-word` has already claimed that part of the string.

So the simplest fix is to move `string` before `conventionally-named-word` or `conventionally-named-word` after `string`. Without going into the details, please prefer the first option.
You also have to add the usual lookbehind to string.

__Some additional information:__

So why does the current (or old) pattern work? Let's look at how Prism matches the above example:

1. It tries to match `comment` and succeeds because the lookbehind is essentially non-existent.   The text is now tokenized as `["\"", ["comment", "! \""]]`.
2. It tries to match `conventionally-named-word` and fails because it's partially claimed by `comment` (first come first served).
3. Prism tries to match `string`. Because `string` is [greedy](https://prismjs.com/extending.html#language-definitions) and the match of `string` starts before the comment match, it will remove the comment highlighting and set it as `string`. The text is now tokenized as `[["string", "\"! \""]]`.

So the reason why it worked is that `comment` prevented `conventionally-named-word` from matching some text before `string` could. The reason why `string` can't remove `conventionally-named-word` is that they start at the same character position in which case it's first come first served again.

Reply:
> `/(?:![\t ].*|!$)/`

`a!` is simply a word named `a!`, and must not be tokenised as a comment. The `!` must be surrounded by whitespace in order for it to be a comment token.

Reply:
Yup. But it is tokenized as a comment right now.

Reply:
The fact that `"!` is matched by `conventionally-named-word` for ending with `!` is a bug. 

Also a bug is that `conventionally-named-word` allows `"` to start a word name; that should be disallowed just like it is in `normal-word`. I'm going to fix  those. 

It's a little bit sad I have to add whitespace lookbehind to `string`, because `"1""2"` is `"1" "2"` in the reference implementation, which is an implementation detail of the special treatment of the builtin string literal. 

The biggest worry is moving `string` before `conventionally-named-word`  -- hopefully it won't be necessary when `conventionally-named-word` correctly disallows words to start with `"`.

Reply:
Fixing `conventionally-named-word` will also work. I guess you mean like `(^|\s)(?!")(?:set-\S+|change-\S+|...)(?=\s|$)`, right?

Reply:
I should have used the negative lookahead at first instead of adding `"` to the classes that needed it, but yes, that's the fix now.